### ISSUE TITLE
TLSRSACertificateKeyFile sanity checks fail unexpectedly for passphrase-protected keys

### ISSUE THREAD

**Issue by gmaxbass** [proftpd/proftpd#1179]
### What I Did
I compiled **proftpd 1.3.7a with openssl-1.1.1g on RedHat 8** ( tried also on CentOS Stream 8 and Oracle Linux 8 with the same behavior)
(added packages on RH8: openssl-devel-1.1.1g, openldap-devel, libsodium-devel)
```
./configure --with-modules=mod_ldap:mod_sftp:mod_tls --enable-openssl --prefix=/usr/local/proftpd-1.3.7a
```
Symbolic link ln -s /usr/local/proftpd-1.3.7a  /usr/local/proftpd-ssl-sftp

**Self signed cert/key with passphrase** created using openssl (1.1.1g)
```
openssl req -newkey rsa:2048 -keyout coll-ftpssrv.key -x509 -days 365 -out coll-ftpssrv.cert
```
Checked the created key file with passphrase without issues with
```
openssl rsa -in coll-ftpssrv.key -text
```

**TLS section in my proftdp.conf:**
```
TLSEngine                       on
TLSRequired                     on
TLSLog                          /usr/local/proftpd-ssl-sftp/var/log/tls.log
TLSProtocol                     TLSv1.2
TLSRSACertificateFile           /usr/local/proftpd-ssl-sftp/etc/certificates/coll-ftpssrv.cert
TLSRSACertificateKeyFile        /usr/local/proftpd-ssl-sftp/etc/certificates/coll-ftpssrv.key
TLSPassPhraseProvider		/usr/local/proftpd-ssl-sftp/etc/startphrase
TLSVerifyClient                 off
```

**Syntax check** with:
```
/usr/local/proftpd-ssl-sftp/sbin/proftpd -c  /usr/local/proftpd-ssl-sftp/etc/proftpd.conf -td10  
Checking syntax of configuration file
2021-02-22 11:35:04,025 coll-ftpssrv proftpd[63632]: using TCP receive buffer size of 87380 bytes
2021-02-22 11:35:04,026 coll-ftpssrv proftpd[63632]: using TCP send buffer size of 16384 bytes
2021-02-22 11:35:04,026 coll-ftpssrv proftpd[63632]: mod_ldap/2.9.5: compiled using LDAP vendor 'OpenLDAP', LDAP API version 3001
2021-02-22 11:35:04,029 coll-ftpssrv proftpd[63632]: mod_ldap/2.9.5 linked with LDAP vendor 'OpenLDAP' (LDAP API version 3001, vendor version 20446), features: X_OPENLDAP
2021-02-22 11:35:04,029 coll-ftpssrv proftpd[63632]: mod_ldap/2.9.5: LDAP TLS package = OpenSSL
2021-02-22 11:35:04,029 coll-ftpssrv proftpd[63632]: mod_sftp/1.0.1: using OpenSSL 1.1.1g FIPS  21 Apr 2020
2021-02-22 11:35:04,030 coll-ftpssrv proftpd[63632]: mod_sftp/1.0.1: using libsodium-1.0.18
2021-02-22 11:35:04,030 coll-ftpssrv proftpd[63632]: mod_tls/2.9: using OpenSSL 1.1.1g FIPS  21 Apr 2020
2021-02-22 11:35:04,031 coll-ftpssrv proftpd[63632]: PathDenyFilter: compiling deny regex '(.ftpaccess)$'
2021-02-22 11:35:04,031 coll-ftpssrv proftpd[63632]: retrieved UID 65534 for user 'nobody'
2021-02-22 11:35:04,031 coll-ftpssrv proftpd[63632]: retrieved GID 65534 for group 'nobody'
2021-02-22 11:35:04,031 coll-ftpssrv proftpd[63632]: <IfModule>: using 'mod_ldap.c' section at line 32
2021-02-22 11:35:04,031 coll-ftpssrv proftpd[63632]: LDAPServer: parsed URL 'ldap://ldap.xxx.it:389/??sub' as 'ldap://ldap.xxx.it:389/??sub'
2021-02-22 11:35:04,032 coll-ftpssrv proftpd[63632]: ROOT PRIVS at mod_tls.c:13195
2021-02-22 11:35:04,032 coll-ftpssrv proftpd[63632]: RELINQUISH PRIVS at mod_tls.c:13197
2021-02-22 11:35:04,032 coll-ftpssrv proftpd[63632]: ROOT PRIVS at mod_tls.c:14339
2021-02-22 11:35:04,033 coll-ftpssrv proftpd[63632]: RELINQUISH PRIVS at mod_tls.c:14342
2021-02-22 11:35:04,033 coll-ftpssrv proftpd[63632]: ROOT PRIVS at mod_tls.c:14364
2021-02-22 11:35:04,035 coll-ftpssrv proftpd[63632]: RELINQUISH PRIVS at mod_tls.c:14381
**_2021-02-22 11:35:04,035 coll-ftpssrv proftpd[63632]: fatal: TLSRSACertificateKeyFile: unable to use '/usr/local/proftpd-ssl-sftp/etc/certificates/coll-ftpssrv.key':
  (1) error:06065064:digital envelope routines:EVP_DecryptFinal_ex:bad decrypt
  (2) error:23077074:PKCS12 routines:PKCS12_pbe_crypt:pkcs12 cipherfinal error
  (3) error:2306A075:PKCS12 routines:PKCS12_item_decrypt_d2i:pkcs12 pbe crypt error
  (4) error:0907B00D:PEM routines:PEM_read_bio_PrivateKey:ASN1 lib
  (5) error:140B0009:SSL routines:SSL_CTX_use_PrivateKey_file:PEM lib on line 105 of '/usr/local/proftpd-ssl-sftp/etc/proftpd.conf'_**
```
I get the same errors if I try to start proftpd in standalone mode.

_If I create a cert/key without passphrase proftpd starts without issues._

### What I Expected/Wanted
I expect proftpd starting with cert/key with passphrase being able to decrypt the key file and starting TLSPassPhraseProvider

_If proftpd is built in the same way on **CentOS_7** cert/key with passphrase it works without any issue._
On Centos_7 openssl version is 1.0.2 .

I tested also with a cert/key with passphrase signed by a certification authority with the same behaviour.

### ProFTPD Version and Configuration
```
**./proftpd -V**
Compile-time Settings:
  Version: 1.3.7a (maint)
  Platform: LINUX [Linux 4.18.0-240.15.1.el8_3.x86_64 x86_64]
  Built: Mon Feb 22 2021 11:18:52 CET
  Built With:
    configure  '--with-modules=mod_ldap:mod_sftp:mod_tls' '--enable-openssl' '--prefix=/usr/local/proftpd-1.3.7a'

  CFLAGS: -g2 -O2 -Wall -fno-omit-frame-pointer
  LDFLAGS: -L$(top_srcdir)/lib -L$(top_builddir)/lib  -rdynamic 
  LIBS:  -lssl -lcrypto -lsodium -lssl  -lldap -llber  -lcrypto -lsupp -lcrypt -ldl  -pthread

  Files:
    Configuration File:
      /usr/local/proftpd-1.3.7a/etc/proftpd.conf
    Pid File:
      /usr/local/proftpd-1.3.7a/var/proftpd.pid
    Scoreboard File:
      /usr/local/proftpd-1.3.7a/var/proftpd.scoreboard

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    - Autoshadow support
    - Controls support
    - curses support
    - Developer support
    - DSO support
    + IPv6 support
    + Largefile support
    - Lastlog support
    - Memcache support
    - ncurses support
    - NLS support
    + OpenSSL support (OpenSSL 1.1.1g FIPS  21 Apr 2020, FIPS enabled)
    - PCRE support
    - POSIX ACL support
    - Redis support
    + Sendfile support
    + Shadow file support
    + Sodium support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10

**./proftpd -l**
Compiled-in modules:
  mod_core.c
  mod_xfer.c
  mod_rlimit.c
  mod_auth_unix.c
  mod_auth_file.c
  mod_auth.c
  mod_ls.c
  mod_log.c
  mod_site.c
  mod_delay.c
  mod_facts.c
  mod_ldap.c
  mod_sftp.c
  mod_tls.c
```
During the make i saw (only in RH8 build, not in CentOS_7) some warning regarding RSA like
```
_kex.c: In function ‘write_kexrsa_pubkey’:
kex.c:3661:26: warning: passing argument 2 of ‘RSA_get0_key’ from incompatible pointer type [-Wincompatible-pointer-types]
   RSA_get0_key(kex->rsa, &rsa_n, &rsa_e, NULL);
                          ^~~~~~
In file included from /usr/include/openssl/x509.h:25,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from kex.c:25:
/usr/include/openssl/rsa.h:210:34: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **n, const BIGNUM **e, const BIGNUM **d);
                   ~~~~~~~~~~~~~~~^
kex.c:3661:34: warning: passing argument 3 of ‘RSA_get0_key’ from incompatible pointer type [-Wincompatible-pointer-types]
   RSA_get0_key(kex->rsa, &rsa_n, &rsa_e, NULL);
                                  ^~~~~~
In file included from /usr/include/openssl/x509.h:25,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from kex.c:25:
/usr/include/openssl/rsa.h:210:52: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **n, const BIGNUM **e, const BIGNUM **d);
                                     ~~~~~~~~~~~~~~~^
kex.c: In function ‘write_kexrsa_done’:
kex.c:3720:26: warning: passing argument 2 of ‘RSA_get0_key’ from incompatible pointer type [-Wincompatible-pointer-types]
   RSA_get0_key(kex->rsa, &rsa_n, &rsa_e, NULL);
                          ^~~~~~
In file included from /usr/include/openssl/x509.h:25,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from kex.c:25:
/usr/include/openssl/rsa.h:210:34: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **n, const BIGNUM **e, const BIGNUM **d);
                   ~~~~~~~~~~~~~~~^
kex.c:3720:34: warning: passing argument 3 of ‘RSA_get0_key’ from incompatible pointer type [-Wincompatible-pointer-types]
   RSA_get0_key(kex->rsa, &rsa_n, &rsa_e, NULL);
                                  ^~~~~~
In file included from /usr/include/openssl/x509.h:25,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from kex.c:25:
/usr/include/openssl/rsa.h:210:52: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **n, const BIGNUM **e, const BIGNUM **d);
                                     ~~~~~~~~~~~~~~~^
gcc  -DHAVE_CONFIG_H  -DLINUX  -I. -I../.. -I../../include -I./../../include  -g2 -O2 -Wall -fno-omit-frame-pointer -c keys.c
keys.c: In function ‘rsa_compare_keys’:
keys.c:1910:28: warning: passing argument 2 of ‘RSA_get0_key’ from incompatible pointer type [-Wincompatible-pointer-types]
   RSA_get0_key(remote_rsa, &remote_rsa_n, &remote_rsa_e, NULL);
                            ^~~~~~~~~~~~~
In file included from /usr/include/openssl/x509.h:25,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from keys.c:25:
/usr/include/openssl/rsa.h:210:34: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **n, const BIGNUM **e, const BIGNUM **d);
                   ~~~~~~~~~~~~~~~^
keys.c:1910:43: warning: passing argument 3 of ‘RSA_get0_key’ from incompatible pointer type [-Wincompatible-pointer-types]
   RSA_get0_key(remote_rsa, &remote_rsa_n, &remote_rsa_e, NULL);
                                           ^~~~~~~~~~~~~
In file included from /usr/include/openssl/x509.h:25,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from keys.c:25:
/usr/include/openssl/rsa.h:210:52: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **n, const BIGNUM **e, const BIGNUM **d);
                                     ~~~~~~~~~~~~~~~^
keys.c:1911:27: warning: passing argument 2 of ‘RSA_get0_key’ from incompatible pointer type [-Wincompatible-pointer-types]
   RSA_get0_key(local_rsa, &local_rsa_n, &local_rsa_e, NULL);
                           ^~~~~~~~~~~~
In file included from /usr/include/openssl/x509.h:25,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from keys.c:25:
/usr/include/openssl/rsa.h:210:34: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **n, const BIGNUM **e, const BIGNUM **d);
                   ~~~~~~~~~~~~~~~^
keys.c:1911:41: warning: passing argument 3 of ‘RSA_get0_key’ from incompatible pointer type [-Wincompatible-pointer-types]
   RSA_get0_key(local_rsa, &local_rsa_n, &local_rsa_e, NULL);
                                         ^~~~~~~~~~~~
In file included from /usr/include/openssl/x509.h:25,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from keys.c:25:
/usr/include/openssl/rsa.h:210:52: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **n, const BIGNUM **e, const BIGNUM **d);
                                     ~~~~~~~~~~~~~~~^
keys.c: In function ‘dsa_compare_keys’:
keys.c:1971:28: warning: passing argument 2 of ‘DSA_get0_pqg’ from incompatible pointer type [-Wincompatible-pointer-types]
   DSA_get0_pqg(remote_dsa, &remote_dsa_p, &remote_dsa_q, &remote_dsa_g);
                            ^~~~~~~~~~~~~
In file included from /usr/include/openssl/x509.h:26,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from keys.c:25:
/usr/include/openssl/dsa.h:178:34: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **p, const BIGNUM **q, const BIGNUM **g);
                   ~~~~~~~~~~~~~~~^
keys.c:1971:43: warning: passing argument 3 of ‘DSA_get0_pqg’ from incompatible pointer type [-Wincompatible-pointer-types]
   DSA_get0_pqg(remote_dsa, &remote_dsa_p, &remote_dsa_q, &remote_dsa_g);
                                           ^~~~~~~~~~~~~
In file included from /usr/include/openssl/x509.h:26,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from keys.c:25:
/usr/include/openssl/dsa.h:178:52: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **p, const BIGNUM **q, const BIGNUM **g);
                                     ~~~~~~~~~~~~~~~^
keys.c:1971:58: warning: passing argument 4 of ‘DSA_get0_pqg’ from incompatible pointer type [-Wincompatible-pointer-types]
   DSA_get0_pqg(remote_dsa, &remote_dsa_p, &remote_dsa_q, &remote_dsa_g);
                                                          ^~~~~~~~~~~~~
In file included from /usr/include/openssl/x509.h:26,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from keys.c:25:
/usr/include/openssl/dsa.h:178:70: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **p, const BIGNUM **q, const BIGNUM **g);
                                                       ~~~~~~~~~~~~~~~^
keys.c:1972:27: warning: passing argument 2 of ‘DSA_get0_pqg’ from incompatible pointer type [-Wincompatible-pointer-types]
   DSA_get0_pqg(local_dsa, &local_dsa_p, &local_dsa_q, &local_dsa_g);
                           ^~~~~~~~~~~~
In file included from /usr/include/openssl/x509.h:26,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from keys.c:25:
/usr/include/openssl/dsa.h:178:34: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **p, const BIGNUM **q, const BIGNUM **g);
                   ~~~~~~~~~~~~~~~^
keys.c:1972:41: warning: passing argument 3 of ‘DSA_get0_pqg’ from incompatible pointer type [-Wincompatible-pointer-types]
   DSA_get0_pqg(local_dsa, &local_dsa_p, &local_dsa_q, &local_dsa_g);
                                         ^~~~~~~~~~~~
In file included from /usr/include/openssl/x509.h:26,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from keys.c:25:
/usr/include/openssl/dsa.h:178:52: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **p, const BIGNUM **q, const BIGNUM **g);
                                     ~~~~~~~~~~~~~~~^
keys.c:1972:55: warning: passing argument 4 of ‘DSA_get0_pqg’ from incompatible pointer type [-Wincompatible-pointer-types]
   DSA_get0_pqg(local_dsa, &local_dsa_p, &local_dsa_q, &local_dsa_g);
                                                       ^~~~~~~~~~~~
In file included from /usr/include/openssl/x509.h:26,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from keys.c:25:
/usr/include/openssl/dsa.h:178:70: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **p, const BIGNUM **q, const BIGNUM **g);
                                                       ~~~~~~~~~~~~~~~^
keys.c:1973:28: warning: passing argument 2 of ‘DSA_get0_key’ from incompatible pointer type [-Wincompatible-pointer-types]
   DSA_get0_key(remote_dsa, &remote_dsa_pub_key, NULL);
                            ^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/openssl/x509.h:26,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from keys.c:25:
/usr/include/openssl/dsa.h:181:34: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **pub_key, const BIGNUM **priv_key);
                   ~~~~~~~~~~~~~~~^~~~~~~
keys.c:1974:27: warning: passing argument 2 of ‘DSA_get0_key’ from incompatible pointer type [-Wincompatible-pointer-types]
   DSA_get0_key(local_dsa, &local_dsa_pub_key, NULL);
                           ^~~~~~~~~~~~~~~~~~
In file included from /usr/include/openssl/x509.h:26,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from keys.c:25:
/usr/include/openssl/dsa.h:181:34: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **pub_key, const BIGNUM **priv_key);
                   ~~~~~~~~~~~~~~~^~~~~~~
keys.c: In function ‘get_openssh_cipher’:
keys.c:2735:24: warning: assignment to ‘const EVP_CIPHER * (*)(void)’ {aka ‘const struct evp_cipher_st * (*)(void)’} from incompatible pointer type ‘const EVP_CIPHER *’ {aka ‘const struct evp_cipher_st *’} [-Wincompatible-pointer-types]
     cipher->get_cipher = sftp_crypto_get_cipher(name, NULL, NULL);
                        ^
keys.c: In function ‘decrypt_openssh_data’:
keys.c:2814:18: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   openssl_cipher = (cipher->get_cipher)();
                  ^
keys.c: In function ‘deserialize_openssh_private_key’:
keys.c:2893:52: warning: unused variable ‘res’ [-Wunused-variable]
     uint32_t public_keylen = 0, secret_keylen = 0, res = 0;
                                                    ^~~
keys.c: In function ‘unwrap_openssh_private_key’:
keys.c:3013:12: warning: variable ‘bufsz’ set but not used [-Wunused-but-set-variable]
   uint32_t bufsz, buflen, kdf_len = 0, key_count = 0, encrypted_len = 0;
            ^~~~~
keys.c: In function ‘get_rsa_hostkey_data’:
keys.c:3505:21: warning: passing argument 2 of ‘RSA_get0_key’ from incompatible pointer type [-Wincompatible-pointer-types]
   RSA_get0_key(rsa, &rsa_n, &rsa_e, NULL);
                     ^~~~~~
In file included from /usr/include/openssl/x509.h:25,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from keys.c:25:
/usr/include/openssl/rsa.h:210:34: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **n, const BIGNUM **e, const BIGNUM **d);
                   ~~~~~~~~~~~~~~~^
keys.c:3505:29: warning: passing argument 3 of ‘RSA_get0_key’ from incompatible pointer type [-Wincompatible-pointer-types]
   RSA_get0_key(rsa, &rsa_n, &rsa_e, NULL);
                             ^~~~~~
In file included from /usr/include/openssl/x509.h:25,
                 from /usr/include/openssl/x509v3.h:14,
                 from mod_sftp.h:80,
                 from keys.c:25:
/usr/include/openssl/rsa.h:210:52: note: expected ‘const BIGNUM **’ {aka ‘const struct bignum_st **’} but argument is of type ‘BIGNUM **’ {aka ‘struct bignum_st **’}
                   const BIGNUM **n, const BIGNUM **e, const BIGNUM **d);
                                     ~~~~~~~~~~~~~~~^_
```

**Comment by Castaglia** [comment]
The compiler warnings about `kex.c`, `keys.c` are for the mod_sftp module, not the mod_tls module.  So while those compiler warnings may be concerning, I don't believe they are relevant to your core issue -- that of `TLSPassPhraseProvider` not working as expected on Centos 8.

I'll see if I can reproduce this behavior locally, using Docker.

**Comment by Castaglia** [comment]
What does `ls -al /usr/local/proftpd-ssl-sftp/etc/startphrase` show?

Also, does the passphrase emitted by that `/usr/local/proftpd-ssl-sftp/etc/startphrase` script end with a newline character, or not?  In theory, it should not; when you provide a passphrase to `openssl req ...`, it trims off the newline.  Thus to match, the `TLSPassPhraseProvider` script needs to write, to `stdout`, the same passhrase _without newline_.  If, for example, you are using a shell script that uses `echo`, then you would want to use `echo -n $PASSPHRASE`, such that the `-n` command-line option to `echo` would omit that trailing newline.

**Comment by gmaxbass** [comment]
Hi,
the "startphrase" script does an echo $passphrase.
You say that an "echo -n" is needed, but this does not explain why in CentOS7 the same script works.

**Comment by Castaglia** [comment]
I'd have to see the full script (without the real passphrase) to say for certain.  It is possible that the shell/interpreter which runs the script behaves differently between Centos 7 and Centos 8; some distros are not good about preserving _full_ backward compatibility of scripts/interpreters.

**Comment by gmaxbass** [comment]
Hi,
As you suggested I created the startphrase script with the following content only:

echo -n "mypasshrase"

I get the same error messages I reported above.
I tried also to change different shells adding #!/usr/bin/bash or #!/usr/bin/sh with the same result.

Thanks

**Comment by Castaglia** [comment]
OK, thanks for the info.

I'm using a `centos:8` Docker image locally, and the latest ProFTPD source code; using your instructions, I'm able to reproduce the reported behavior.  And, for comparison, when I use the same OpenSSL version and same ProFTPD source code on my Mac, it works.  Very odd.  But now that I can reproduce this behavior, I can track down the causes.

**Comment by Castaglia** [comment]
I think that https://github.com/proftpd/proftpd/commit/00e3acf3b1e23bac593e314025bb657264e68795 is the cause; it added some sanity checks to things including, among others, the `TLSRSACertificateKeyFile` directive.  (In my local regression testsuite, I have my RSA certificate and key in the same file, and thus don't use `TLSRSACertificateKeyFile` -- and that's why I didn't see this behavior locally.)

I'm working on a fix now.

**Comment by Castaglia** [comment]
This should now be fixed in master, and backported to the 1.3.7 branch.  Thanks!
